// containers/leftarea/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'

import Path from 'components/path'
import FirstPage from './FirstPage'
import SecondPage from './SecondPage'
import ThirdPage from './ThirdPage'
import FourthPage from './FourthPage'

class LeftArea extends Component {
   constructor (props) {
      super (props)
      this.next = this.next.bind (this)
      this.toPage = this.toPage.bind (this)
      this.state = {page: 1, openedPage: 0}
   }

   next (val) {
      const {addparam} = this.props.state

      const P1 = [1, 2, 3, 5, 6, 7].find (el => (addparam[el] && addparam[el].visible === false))
      const P2 = [10, 11, 12, 14, 15, 16].find (el => (addparam[el] && addparam[el].visible === false))
      const P3 = [21, 22, 23, 24].find (el => (addparam[el] && addparam[el].visible === false))

      if (val === 1 && P1 === undefined) {
         this.setState ({
            page: 2,
            openedPage: 1
         })
      }
      if (val === 2 && P2 === undefined) {
         this.setState ({
            page: 3,
            openedPage: 2
         })
      }
      if (val === 3 && P3 === undefined) {
         this.setState ({
            page: 4,
            openedPage: 3
         })
      }
   }

   toPage (val) {
      if (val <= this.state.openedPage + 1) {
         this.setState ({
            page: val
         })
      }
   }

   render () {
      return (
         <div>
            <Path parentState={this.state} toPage={this.toPage}/>
            {this.state.page === 1 && <FirstPage next={this.next}/>}
            {this.state.page === 2 && <SecondPage next={this.next}/>}
            {this.state.page === 3 && <ThirdPage next={this.next}/>}
            {this.state.page === 4 && <FourthPage/>}
         </div>
      )
   }
}

const mapStateToProps = state => ({state})

export default connect (mapStateToProps) (LeftArea)
