// containers/leftarea/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'

import FirstPage from './FirstPage'
import SecondPage from './SecondPage'
import ThirdPage from './ThirdPage'
import FourthPage from './FourthPage'

class LeftArea extends Component {
   constructor (props) {
      super (props)
      this.next = this.next.bind (this)
      this.state = {
         page1: true,
         page2: false,
         page3: false,
         page4: false
      }
   }

   next (val) {
      const {addparam} = this.props.state

      const P1 = [1, 2, 3, 5, 6, 7].find (el => (addparam[el] && addparam[el].visible === false))
      const P2 = [10, 11, 12, 14, 15, 16].find (el => (addparam[el] && addparam[el].visible === false))
      const P3 = [21, 22, 23, 24].find (el => (addparam[el] && addparam[el].visible === false))

      if (val === 1 && P1 === undefined) {
         this.setState ({
            page1: false,
            page2: true
         })
      }
      if (val === 2 && P2 === undefined) {
         this.setState ({
            page2: false,
            page3: true
         })
      }
      if (val === 3 && P3 === undefined) {
         this.setState ({
            page3: false,
            page4: true
         })
      }
   }

   render () {
      return (
         <div>
            {this.state.page1 && <FirstPage next={this.next}/>}
            {this.state.page2 && <SecondPage next={this.next}/>}
            {this.state.page3 && <ThirdPage next={this.next}/>}
            {this.state.page4 && <FourthPage/>}
         </div>
      )
   }
}

const mapStateToProps = state => ({state})

export default connect (mapStateToProps) (LeftArea)
