// containers/leftarea/SecondPage.js
import React, {Component} from 'react'
import Combo from 'components/combo'
import {connect} from 'react-redux'
import {addPARAM} from 'action'

let allAreaFullRoot = false

class SecondPage extends Component {
   constructor (props) {
      super (props)
      this.state = {
         checkAllEmptyPlaces: false
      }
   }

   onContinue () {
      this.setState ({checkAllEmptyPlaces: true})
      this.props.next (2)
   }

   onNextPage () {
      if (allAreaFullRoot) this.props.addPARAM ("Первый пройден")
   }


   render () {
      return (

         <div>
            <h2 className="blue">Billing info</h2>
            <h4 className="blue">Billing Contact</h4>
            <Combo
               id='10'
               type="fio"
               placeholder="Full Name"
               count="6"
               checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
               alt="Please enter billing full name"
            />

            <Combo
               id='11'
               type="email"
               placeholder="Email Address"
               checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
               alt="Please enter billing Email Address"
            />

            <h4 className="blue">Billing Address</h4>
            <Combo
               id='12'
               type="fio"
               placeholder="Street Address"
               checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
               alt="Please enter billing Street Address"
            />
            <Combo
               type="fio"
               placeholder="Apt, Suite, Bldg, Gate Code. (optional)"
            />
            <Combo
               id='14'
               type="fio"
               placeholder="City"
               checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
               alt="Please enter billing City"
            />
            <div>
               <Combo
                  id='15'
                  width="150px"
                  placeholder="Country"
                  checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
                  alt="Please enter billing Country"
               />
               <Combo
                  id='16'
                  width="150px"
                  placeholder="ZIP"
                  checkAllEmptyPlaces={this.state.checkAllEmptyPlaces}
                  alt="Please enter billing ZIP"
               />
            </div>

            <div
               className="button"
               onMouseDown={this.onContinue.bind (this)}
               onMouseUp={this.onNextPage.bind (this)}>
               Continue
            </div>
         </div>
      )
   }
}

const mapDispatchWithProps = {addPARAM}

export default connect (null, mapDispatchWithProps) (SecondPage)
