// action/index.js
import {
   FETCH_COMBO_START,
   FETCH_COMBO_SUCCESS,
   FETCH_COMBO_FAILURE,

   FETCH_COUNTRY_START,
   FETCH_COUNTRY_SUCCESS,
   FETCH_COUNTRY_FAILURE,

   FETCH_ADD_PARAM_START,
   FETCH_ADD_PARAM_SUCCESS,
   FETCH_ADD_PARAM_FAILURE,
} from 'actionTypes'
import {changeCombo as changeComboApi, changeCountryApi} from 'api'

export const changeCombo = (val, type, count) => async dispatch => {
   if (val === "" || count === 0) {
      return dispatch ({type: FETCH_COMBO_SUCCESS, payload: []})
   }

   dispatch ({type: FETCH_COMBO_START})
   try {
      let datas = await changeComboApi (val, type, count);
      dispatch ({
         type: FETCH_COMBO_SUCCESS,
         payload: datas.data.suggestions
      })
   } catch (err) {
      dispatch ({
         type: FETCH_COMBO_FAILURE,
         payload: err,
         error: true
      })
   }

}

export const changeCountry = (val) => async dispatch => {
   if (val === "") {
      return dispatch ({type: FETCH_COUNTRY_SUCCESS, payload: []})
   }

   dispatch ({type: FETCH_COUNTRY_START})
   try {
      let datas = await changeCountryApi (val);
      dispatch ({
         type: FETCH_COUNTRY_SUCCESS,
         payload: datas
      })
   } catch (err) {
      dispatch ({
         type: FETCH_COUNTRY_FAILURE,
         payload: err,
         error: true
      })
   }

}


export const addPARAM = (params) => async dispatch => {

   dispatch ({type: FETCH_ADD_PARAM_START})
   try {
      dispatch ({
         type: FETCH_ADD_PARAM_SUCCESS,
         payload: params
      })
   } catch (err) {
      dispatch ({
         type: FETCH_ADD_PARAM_FAILURE,
         payload: err,
         error: true
      })
   }

}
