// components/combo/index.js
/**
 *  type {String} [fio|address|email]
 */
/** count {Number} количество подсказок, если ноль не будет подсказок*/

import './index.css'

import React, {Component} from 'react'
import {connect} from 'react-redux'
import {changeCombo, changeCountry, addPARAM} from 'action'
import {getGeolocationCity} from "api";


class Combo extends Component {

   constructor (props) {
      super (props)
      let empty = !!this.props.state.addparam[this.props.id]
      this.state = {
         active: false,
         changedEl: -1, // выбранный вариант подсказки
         emptyRoot: empty // если false требует заполнения
      }
   }

   componentDidMount () {
      var val = this.props.state.addparam[this.props.id]
      this.refs.inp.value = val ? val.value : "";
      if (this.props.setValue) {
         this.setValue = this.props.setValue
      }
   }

   onInputChange (val) {

      if (this.props.extCountry) {
         this.props.changeCountry (
            (val == null) ? '' : this.refs.inp.value,
            this.props.type,
            (this.props.count) ? +this.props.count : 4,
            this.props.extCountry // если запросили Country
         )
      } else {
         this.props.changeCombo (
            (val == null) ? '' : this.refs.inp.value,
            this.props.type,
            (this.props.count) ? +this.props.count : 4
         )
      }

      this.setState ({emptyRoot: this.refs.inp.value.length > 2})
      this.addParam (this.refs.inp.value, this.refs.inp.value.length > 2)
      this.onVerification ()
   }

   onVerification () {
      var value = this.refs.inp.value
      var rep = new RegExp ("[" + this.props.exept + "]")
      if (rep.test (value)) {
         value = value.replace (rep, '')
         this.refs.inp.value = value
      }
   }

   onMouseLeave () {
      this.onInputChange (null)
   }

   onFocus () {
      this.setState ({active: true})
   }

   onBlur () {
      this.setState ({active: false})
   }

   onChose (val) {
      this.refs.inp.value = val
      this.onInputChange (null)
      this.addParam (val)
   }

   addParam (value, visible) {
      this.props.addPARAM ([{
         id: this.props.id,
         visible: visible || this.state.emptyRoot,
         placeholder: this.props.placeholder,
         value
      }])
   }

   /**
    * выбраем вариант на клавиатуре
    */
   onKeyDown (event) {

      if (event.keyCode === 40) {
         let changed = ++this.state.changedEl;
         if (changed > +this.props.count - 1) changed = +this.props.count - 1
         this.setState ({changedEl: changed})
      }
      if (event.keyCode === 38) {
         let changed = --this.state.changedEl;
         if (changed < -1) {
            changed = -1
            if (this.props.extCountry) {
               changed = this.props.state.combo.length - 1
               this.setState ({changedEl: changed})
            }
         }

         this.setState ({changedEl: changed})
      }
      if (event.keyCode === 13) {
         this.onChose (this.props.state.combo[this.state.changedEl])
         this.setState ({changedEl: -1})
      }
   }

   onMouseOverDown () {
      this.setState ({changedEl: -1})
   }

   getGeolocation () {
      const makeRequest = async () => {
         let res = await getGeolocationCity ()
         this.refs.inp.value = res;
      }
      makeRequest ()
   }

   render () {
      const {combo} = this.props.state
      let visibleDownClass = (this.state.active && combo.length > 0) ? "combobox-down" : "combobox-down combobox-down-hide";
      let divStyle = {width: this.props.width}
      let altVisibleStyle = {display: "none"}

      let emptyClass = {}

      if (this.props.checkAllEmptyPlaces && !this.state.emptyRoot) {
         emptyClass = {background: "lightyellow"}
      }

      if (emptyClass.hasOwnProperty ("background") && !!this.props.alt && this.state.active && combo.length === 0) {
         altVisibleStyle = {display: "block"}
      }

      if (this.props.extCountry) {
         visibleDownClass = visibleDownClass.replace ("combobox-down", "combobox-up")
      }

      return (
         <div
            onKeyDown={this.onKeyDown.bind (this)}
            onFocus={this.onFocus.bind (this)}
            style={divStyle}
            className="combobox"
            onBlur={this.onBlur.bind (this)}
            onMouseLeave={this.onMouseLeave.bind (this)}
         >
            <input
               style={emptyClass}
               className="combobox-input"
               ref="inp"
               type='text'
               onChange={this.onInputChange.bind (this)}
               placeholder={this.props.placeholder}
            />
            <div className={visibleDownClass}>
               {
                  this.state.active &&
                  combo &&
                  combo.map &&
                  combo.map ((el, index) => {
                     let buttonClass = this.state.changedEl === index ? 'combobox-down-active' : '';
                     return (
                        <div
                           key={index}
                           className={buttonClass}
                           onMouseOver={this.onMouseOverDown.bind (this, el)}
                           onMouseDown={this.onChose.bind (this, el)}
                        >{el}</div>
                     )
                  })}
            </div>

            <div
               style={altVisibleStyle}
               className='combobox-alt'>
               {this.props.alt}
            </div>

            {this.props.geolocation &&
            <div className="geolocationButton">
               <img
                  onClick={this.getGeolocation.bind (this)}
                  src="images/geolocation.png"/>
            </div>
            }

         </div>
      )
   }
}

const mapStateToProps = state => ({state})
const mapDispatchWithProps = {changeCombo, changeCountry, addPARAM}


export default connect (mapStateToProps, mapDispatchWithProps) (Combo)

//
// Combo.propTypes = {
//    type: React.PropTypes.string,
//    count: React.PropTypes.string,
//    placeholder: React.PropTypes.string,
// };
